### PR TITLE
refactor tests and plugin infrastructure 

### PR DESCRIPTION
**Pull request by chassing**
* plugins return a list of `GeneratedFiles`
* restructured tests in `test/plugins`. Usage of pytest parametrize feature to avoid code duplications and it's easier to extend for future plugins

### CODE REVIEW

**Comment by fishi0x01** [comment]
First of all: Great refactor on the test suite!!! This is super neat! :heart: :smiley: 

Concerning generating the `__init__.py`, I have concerns. It enforces an opinionated way of how `__init__.py` must look like. But what if a user wants to actually have content in the `__init__.py`?

Additionally, nested dir structures will still suffer from this issue.
E.g. lets assume this nested dir structure:

```bash
.
└── a
    └── b
        └── query.gql
```

This will require the following `__init__.py`s:

```bash
.
└── a
    ├── b
    │   ├── __init__.py
    │   └── query.gql
    └── __init__.py
```

Every directory on the way to the query.gql needs an `__init__.py`.

I.e., even with this change, it won't necessarily save the user from tackling `__init__.py` issues and on top it will introduce an opinionated way of how an `__init__.py` must look like, limiting the user in design choices.

That being said, for our current use-case it is fine now. But long-term I think it would be better to have this as an optional feature flag (but that could be added in a later iteration).

What are your thoughts on this?

File: tests/plugins/test_generate.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+@pytest.mark.parametrize(
+    "query",
+    [
+        # simple query
+        "saas_file_simple",
+        # interface query
+        "complex_inline_fragments_and_collisions",
+        # interface humongous query
+        "saas_humongous",
+        # difficult attribute name query
```
Comment (fishi0x01) [inline tests/plugins/test_generate.py:15]: nit: those comments feel redundant. Ideally, the filename already describes the case properly

File: tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import json
+from pathlib import Path
+import pytest
+
+
+@pytest.fixture()
+def expected_files():
+    """Read fixture files from test/queries/expected/PLUGIN/QUERY and remove .txt suffix."""
+    from qenerate.core.plugin import GeneratedFile
```
Comment (fishi0x01) [inline tests/conftest.py:9]: out of curiosity: why is this inlined instead of declared on top?

File: tests/plugins/test_generate.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        "difficult_attribute_name",
+        # inline fragments
+        "ocp_with_inline_fragments",
+        # json query
+        "saas_file_json",
+    ],
+)
+@pytest.mark.parametrize("plugin_name", plugins.keys())
+def test_generate(schema_raw, expected_files, query, plugin_name):
+    """Test code generation for each QUERY x PLUGIN combinations."""
```
Comment (fishi0x01) [inline tests/plugins/test_generate.py:25]: neat! :tada:

File: tests/plugins/test_generate.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    generated_files = plugin.generate(query_file=query_file, raw_schema=schema_raw)
+    assert generated_files == expected_files(plugin=plugin_name, query=query)
+
+
+@pytest.mark.parametrize(
+    "query, exception",
+    [
+        # anonymous query
+        ["anonymous_query", AnonymousQueryError],
+        # invalid query
```
Comment (fishi0x01) [inline tests/plugins/test_generate.py:37]: nit: same as above - comments feel redundant

File: tests/conftest.py (hunk 2 above)
Comment (chassing) [reply tests/conftest.py:9]: I sometimes face import errors, but that's not the case here.

File: tests/plugins/test_generate.py (hunk 3 above)
Comment (chassing) [reply tests/plugins/test_generate.py:19]: I like this pytest feature so much :)

**Comment by chassing** [comment]
You're right; the `__init__.py` topic is complex, and having a feature flag that gives more control would be great. I'm thinking about the possibility to don't generating a particular file so that the user can adapt it manually after the first generation.

Regarding extensibility by the users, I think qenerate should provide a way to allow it or provide some best practices at least. Maybe we can use `__init__.py` to import a user-defined python file. E.g.

```python
try:
    from .xxxx import *
except ImportError:
    pass
```

**Comment by chassing** [comment]
As discussed, I removed the `__init__.py` generation and will put that into a-i.

**Review by fishi0x01 — APPROVED** [review]
Awesome refactor! :heart:

**Pull request merged**